src/native/directory_traversal.c

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "crt.h"
+#include "java_class_ids.h"
+#include "retry_utils.h"
+#include <aws/common/file.h>
+#include <aws/common/string.h>
+#include <jni.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#        pragma warning(disable : 4221)
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+struct directory_traversal_callback_ctx {
+    JNIEnv *env;
+    jobject handler;
+};
+
+static bool s_on_directory_entry(const struct aws_directory_entry *entry, void *user_data) {
+
+    struct directory_traversal_callback_ctx *ctx = user_data;
+    JNIEnv *env = ctx->env;
+
+    jstring path = aws_jni_string_from_cursor(env, &entry->path);
+    jstring relativePath = aws_jni_string_from_cursor(env, &entry->relative_path);
```

Comment:
I was going to say we should check that these succeeded

BUT 

looks like it can only fail if we're out-of-memory
and we don't deal with that anymore
so maybe add an AWS_FATAL_ASSERT to aws_jni_string_from_cursor()
and add a comment in crt.h that the function cannot fail

Reply:
Changed as suggested, thanks!